src/params.rs

```diff
@@ -0,0 +1,142 @@
+//! Fixed size parameters list with optional subparameters.
+
+use core::fmt::{self, Debug, Formatter};
+
+pub(crate) const MAX_PARAMS: usize = 32;
+
+#[derive(Default)]
+pub struct Params {
+    /// Number of subparameters for each parameter.
+    ///
+    /// For each entry in the `params` slice, this stores the length of the param as number of
+    /// subparams at the same index as the param in the `params` slice.
+    ///
+    /// At the subparam positions the length will always be `0`.
+    subparams: [u8; MAX_PARAMS],
+
+    /// All parameters and subparameters.
+    params: [i64; MAX_PARAMS],
+
+    /// Number of suparameters in the current parameter.
+    current_subparams: u8,
+
+    /// Total number of parameters and subparameters.
+    len: usize,
+}
+
+impl Params {
+    /// Returns the number of parameters.
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns `true` if there are no parameters present.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    /// Returns an iterator over all parameters and subparameters.
+    #[inline]
+    pub fn iter(&self) -> ParamsIter<'_> {
+        ParamsIter::new(self)
+    }
+
+    /// Returns `true` if there is no more space for additional parameters.
+    #[inline]
+    pub(crate) fn is_full(&self) -> bool {
+        self.len == MAX_PARAMS
+    }
+
+    /// Clear all parameters.
+    #[inline]
+    pub(crate) fn clear(&mut self) {
+        self.current_subparams = 0;
+        self.len = 0;
+    }
+
+    /// Add an additional parameter.
+    #[inline]
+    pub(crate) fn push(&mut self, item: i64) {
+        self.subparams[self.len - self.current_subparams as usize] = self.current_subparams + 1;
+        self.params[self.len] = item;
+        self.current_subparams = 0;
+        self.len += 1;
+    }
+
+    /// Add an additional subparameter to the current parameter.
+    #[inline]
+    pub(crate) fn extend(&mut self, item: i64) {
+        self.params[self.len] = item;
+        self.current_subparams += 1;
+        self.len += 1;
+    }
+}
+
+impl<'a> IntoIterator for &'a Params {
+    type IntoIter = ParamsIter<'a>;
+    type Item = &'a [i64];
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+/// Immutable subparameter iterator.
+pub struct ParamsIter<'a> {
+    params: &'a Params,
+    index: usize,
+}
```

Comment:
I wonder if it should be `std::iter::Peekable` iterator? I do see that we don't really need it in alacritty, but it could be useful in general or you think that there's no much value in it?

Reply:
You can peek any iterator by calling `peekable` on it. So no, just a generic `Iterator` should be better I think.